jenkins/python-setup.sh

```diff
@@ -17,10 +17,22 @@
 # chosen over any installed ones, and it adds `/usr/sbin` to the end of the
 # path for the `ip` command, used by pbench-register-tool.
 
-PATH=$(python3 -m site --user-base)/bin:${PATH}:/usr/sbin
+function _pbench_pathins() {
+    if [[ ":${PATH:-}:" != *":${1}:"* ]]; then
+        PATH="${1}${PATH:+:${PATH}}"
+    fi
+}
+
+_pbench_pathins "/usr/sbin"
+_pbench_pathins "/usr/bin"
+_pbench_pathins "$(python3 -m site --user-base)/bin"
+unset -f _pbench_pathins
+export PATH
+
 unset PYTHONPATH
-pip3 install --user -r lint-requirements.txt -r docs/requirements.txt -r agent/requirements.txt -r server/requirements.txt -r agent/test-requirements.txt -r server/test-requirements.txt
+pip3 install --user --no-cache-dir -r lint-requirements.txt -r docs/requirements.txt -r agent/requirements.txt -r server/requirements.txt -r agent/test-requirements.txt -r server/test-requirements.txt
 python3 setup.py develop --user
+export PYTHONPATH=$(python3 -c "import sys; print(':'.join(x for x in sys.path if x))")
 
 # We delete any python compiler caches to avoid conflicts with the python compiler
 # in the container.
```

Comment:
The `find` command on line 39 is ill-timed:  when this script is run, we're typically already inside the container, and we've already run Python several of times (lines 28, 33, 34, and 35), so we've already started creating cached Python compilations.

We really should do this earlier, and we should do it again after the end of the test run.  I suggest moving this to `jenkins/run-unittests` and doing it before we run `python-setup.sh` and again after `tox` finishes.

Reply:
Also, this seems to be intended to perform the same function as `utils/detox`.  Perhaps we should be using that script instead of reproducing the code here (note that the two are using slightly different criteria...).  Also, we're already running that script (via Tox) before invoking `python-setup.sh`, so perhaps we don't need this code here at all (although, I _do_ think it would be a good idea to clear these caches after the test run before we exit the container).

Reply:
Tox appears to offer a [`commands_post`](https://tox.wiki/en/latest/config.html#conf-commands_post) configurator.  Perhaps that would be an appropriate way to invoke `detox` (rather than doing it from either `python-setup.sh` or `jenkins/run-unittests`).  (It would be nicely symmetrical with the existing use of `commands_pre`.)

Reply:
Okay, so this comment prompted a very deep rabbit hole investigation.

We found during this investigation that `utils/detox` is not actually doing anything substantially useful prior to this PR.  The reason is that if you look at the `tox.ini` file, and the CI job's `jenkins/run-pytests` (prior to this PR) you'll notice that we run the `pytests` specifying files in the source tree.  So all tests then run in the source tree, creating `__pycache__` directories all over it.  While there are `__pycache__` directories under the tox virtualenv directories, they are not for our code.

The newly added commit (is/will be) titled, "detox tox without detox", addresses all this by doing the following:

 1. Properly installing our source into the virtualenv
 2. Running the tests by the module path reference so that they execute out of the installed sources
 3. Working to ensure no other files are created in the source tree

This removes the need for `utils/detox` at all.

Reply:
OK, so I'm trying to close the loop on this....  The plan is to install our code into a virtualenv, run the tests out of there, and then dispose of the virtualenv when we're done...which all sounds great, when we have a virtualenv.  

However, if we're running in the Jenkins container, then, as you've said elsewhere, "the container _is_ the virtualenv", but...in that environment I think we're running the tests out of the source tree, and the source tree is a mapped volume such that any Python cached compilation files end up being persistent...so we need something like detox running at the end of the test...I think....  Help?

Reply:
See this [other comment](https://github.com/distributed-system-analysis/pbench/pull/2551/files#r824224828).